src/vanilla.ts

```diff
@@ -107,23 +107,50 @@ const buildProxyFunction = (
       : Object.create(Object.getPrototypeOf(target))
     markToTrack(snap, true) // mark to track
     snapCache.set(target, [version, snap])
-    Reflect.ownKeys(target).forEach((key) => {
+
+    function getValue(key: PropertyKey): any {
       const value = Reflect.get(target, key)
       if (refSet.has(value as object)) {
         markToTrack(value as object, false) // mark not to track
-        snap[key] = value
-      } else if (value instanceof Promise) {
+        return value
+      } else if (proxyStateMap.has(value as object)) {
+        return snapshot(value as object, handlePromise)
+      } else {
+        return value
+      }
+    }
+
+    // This will copy all object getters, functions, and values
+    Reflect.ownKeys(target).forEach((key) => {
+      const value = getValue(key)
+      if (value instanceof Promise) {
         Object.defineProperty(snap, key, {
           get() {
             return handlePromise(value)
           },
         })
-      } else if (proxyStateMap.has(value as object)) {
-        snap[key] = snapshot(value as object, handlePromise)
       } else {
         snap[key] = value
       }
     })
+
+    // Now look for proto getters to essentially cache them
+    const protoGetters = findPrototypeGetters(target)
+    protoGetters.forEach((key) => {
+      const value = getValue(key)
+      if (value instanceof Promise) {
+        Object.defineProperty(snap, key, {
+          get() {
+            return handlePromise(value)
+          },
+        })
+      } else {
+        // We have to use `defineProperty` instead of `snap[key] = value` to explicitly
+        // set the value on the snap instance itself, and not just invoke the setter
+        Object.defineProperty(snap, key, { value })
```

Comment:
Fwiw we could get rid of this special casing if line 133's `snap[key] = value` was _always_ `Object.defineProperty`, but I wasn't sure if that would be desirable or not...

Like, for the non-class-getter fields, is `snap[key] = value` faster/better than `Object.defineProperty`? If not, I can simplify this code by using `Object.defineProperty` for everything, I just wasn't sure...